File: 2. Clean Architecture Objective-C/Unit Tests.xctemplate/___FILEBASENAME___PresenterTests.m
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  Copyright (c) ___YEAR___ ___ORGANIZATIONNAME___. All rights reserved.
//

@import XCTest;
#import "___VARIABLE_sceneName:identifier___Presenter.h"
#import "___VARIABLE_sceneName:identifier___BoundaryModels.h"

#pragma mark Mock
NS_ASSUME_NONNULL_BEGIN
@interface ___VARIABLE_sceneName:identifier___PresenterOutputMock: NSObject <___VARIABLE_sceneName:identifier___PresenterOutput>
@property (nonatomic, nullable) ___VARIABLE_sceneName:identifier___ResponseViewModel *receivedViewModel;
@end

@implementation ___VARIABLE_sceneName:identifier___PresenterOutputMock

- (void)displaySomethingWithViewModel:(___VARIABLE_sceneName:identifier___ResponseViewModel *)viewModel {
    self.receivedViewModel = viewModel;
}

@end
NS_ASSUME_NONNULL_END

#pragma mark -
#pragma mark Test Class
@interface ___VARIABLE_sceneName:identifier___PresenterTests : XCTestCase
@property (nonatomic, nonnull) ___VARIABLE_sceneName:identifier___Presenter *presenter;
@property (nonatomic, nonnull) ___VARIABLE_sceneName:identifier___PresenterOutputMock *output;
@end

@implementation ___VARIABLE_sceneName:identifier___PresenterTests
#pragma mark Test lifecycle
- (void)setUp {
    [super setUp];
    [self setup___VARIABLE_sceneName:identifier___Presenter];
}

- (void)tearDown {
    [super tearDown];
}

#pragma mark Test setup
- (void)setup___VARIABLE_sceneName:identifier___Presenter {
    self.presenter = [___VARIABLE_sceneName:identifier___Presenter new];
    self.output = [___VARIABLE_sceneName:identifier___PresenterOutputMock new];
    self.presenter.output = self.output;
}

#pragma mark Tests
- (void)testPresentSomething {
    // Given
    ___VARIABLE_sceneName:identifier___Response *response = [[___VARIABLE_sceneName:identifier___Response alloc] init];
    
    // When
    [self.presenter presentSomethingWithResponse:response];
    
    // Then
    // Add expectations for output.receivedViewModel here.
}

@end

File: 1. Clean Architecture Swift/Unit Tests.xctemplate/___FILEBASENAME___PresenterTests.swift
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  Copyright (c) ___YEAR___ ___ORGANIZATIONNAME___. All rights reserved.
//

@testable import ___PROJECTNAMEASIDENTIFIER___
import XCTest

class ___VARIABLE_sceneName:identifier___PresenterTests: XCTestCase {
    // MARK: Subject under test
    var presenter: ___VARIABLE_sceneName:identifier___Presenter!
    
    // MARK: Mock
    class OutputMock: ___VARIABLE_sceneName:identifier___PresenterOutput {
        var receivedViewModel: ___VARIABLE_sceneName:identifier___.Response.ViewModel?
        
        func displaySomething(_ viewModel: ___VARIABLE_sceneName:identifier___.Response.ViewModel) {
            receivedViewModel = viewModel
        }
    }
    
    var output: OutputMock!
    
    // MARK: Test lifecycle
    override func setUp() {
        super.setUp()
        setup___VARIABLE_sceneName:identifier___Presenter()
    }
    
    override func tearDown() {
        super.tearDown()
    }
    
    // MARK: Test setup
    func setup___VARIABLE_sceneName:identifier___Presenter() {
        presenter = ___VARIABLE_sceneName:identifier___Presenter()
        
        output = OutputMock()
        presenter.output = output
    }
    
    // MARK: Tests
    func testPresentSomething() {
        // Given
        let response = ___VARIABLE_sceneName:identifier___.Response()
        
        // When
        presenter.presentSomething(response)
        
        // Then
        // Add expectations for output.receivedViewModel here.
    }
}
